### PR TITLE
Feat ICP Context contract

### PR DESCRIPTION
**Pull request by alenmestrov**
# Created ICP Version of Context Contract

## Summary
This PR adds the Internet Computer (ICP) implementation of the context-config contract, mirroring the existing NEAR implementation.

#### New Implementation
- Added ICP-specific contract implementation using Candid types.
- Implemented core types matching NEAR's structure:
  - Base `ICIdentity` wrapper for 32-byte arrays.
  - Identity types: `ICSignerId`, `ICContextIdentity`, `ICContextId`.
  - Application types: `ICApplication`, `ICApplicationId`, `ICBlobId`.
  - Request types: `Request`, `ContextRequest`, `ContextRequestKind`.
  - Capability management: `ICCapability`.

#### Integration Tests
- Added comprehensive test suite covering:
  - Basic context operations.
  - Member management.
  - Capability management.
  - Edge cases and timestamp scenarios.
  - Concurrent operations.

#### Compatibility
- Ensures compatibility with NEAR implementation through proper type conversions.
- Maintains consistent behavior across both platforms.

### Testing
- All tests pass in the local ICP test environment (PocketIC).

### Build and Test Instructions
1. **Prerequisites**: 

- Ensure that [DFX](https://internetcomputer.org/docs/current/developer-docs/setup/install/) is installed.

- Add candid-extractor `cargo add candid-extractor` for generation of .did file

2. **Build Script**: Use the provided `build.sh` script to build, deploy, and prepare the contract for local testing.
   - The script performs the following steps:

      -    Checks that device has wasm32 target installed
      -    Builds the wasm contract
      -    Generates .did file based on contract
      -    Stops the local environment (if its not running it doesn't do anything)
      -    Starts the local environment
      -    Deploys the contract

   

3. **Run Tests**: Execute the following command to build and run the integration tests:
   ```bash
   ./build.sh && cargo test --test integration -- --nocapture
   ```

### Notes

- To be able to test the contract you will need to download and add [pocket-ic](https://internetcomputer.org/docs/current/developer-docs/smart-contracts/test/pocket-ic) in the `context-config` folder 
- Currently, deployment and update of the proxy contract is not implemented, this will be done after finalization of Proxy Contract

### CODE REVIEW

File: contracts/icp/context-config/.env
Comment (miraclx) [inline contracts/icp/context-config/.env:1]: should this be `.gitignore`d?

File: contracts/icp/context-config/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    
+    #[error("Internal error: {0}")]
+    Internal(String),
+}
+
+impl From<ContractError> for String {
+    fn from(error: ContractError) -> Self {
+        error.to_string()
+    }
+}
```
Comment (miraclx) [inline contracts/icp/context-config/src/error.rs]: why not just call `.to_string()` where this is needed?

File: contracts/icp/context-config/src/types.rs [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    fn from(value: SignerId) -> Self {
+        ICSignerId(Identity(value.as_bytes()))
+    }
+}
+
+impl From<Repr<SignerId>> for ICSignerId {
+    fn from(value: Repr<SignerId>) -> Self {
+        ICSignerId(Identity(value.as_bytes()))
+    }
+}
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]: no `U: From<Repr<T>>` needed if `U: From<T>`

```suggestion
```

you can remove the others as well

File: contracts/icp/context-config/src/types.rs [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+impl From<ICSignerId> for SignerId {
+    fn from(value: ICSignerId) -> Self {
+        value.0 .0.rt().expect("Infallible conversion")
+    }
+}
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]: consider implementing `ReprBytes` for all these types?

so these `From` impls would no longer be necessary

File: contracts/icp/context-config/src/types.rs [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+        // Get the signer's public key from the payload
+        let signer_id = f(&self.payload);
+
+        // Convert signer_id to VerifyingKey (public key)
+        let verifying_key =
+            VerifyingKey::from_bytes(&signer_id.0 .0).map_err(|_| "invalid public key")?;
+
+        // Serialize the payload for verification
+        let message =
+            candid::encode_one(&self.payload).map_err(|_| "failed to serialize payload")?;
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]: how about we keep this matching the NEAR one, decode when parsing

**Comment by frdomovic** [comment]
Add this requirement for building and deploying the contract - https://crates.io/crates/candid-extractor

File: contracts/icp/context-config/.env
Comment (alenmestrov) [reply contracts/icp/context-config/.env:1]: this .env file is used by ICP itself. It was automatically created during the project initialization and I didn't touched it at all so we can say that its something like example env

File: contracts/icp/context-config/Cargo.lock
Comment (miraclx) [inline contracts/icp/context-config/Cargo.lock:1]: icp/context-config is already part of the root manifest, so we shouldn't need this here

File: contracts/icp/context-config/src/guard.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+use candid::CandidType;
+use serde::{Deserialize, Serialize};
+
+use crate::types::ICSignerId;
+
+#[derive(CandidType, Serialize, Deserialize, Clone, Debug)]
+pub struct Guard<T> {
+    inner: T,
+    revision: Revision,
+    privileged: Vec<ICSignerId>,
```
Comment (miraclx) [inline contracts/icp/context-config/src/guard.rs]: `BTreeSet`

File: contracts/icp/context-config/src/guard.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    pub fn grant(&mut self, signer_id: ICSignerId) {
+        if !self.inner.contains(&signer_id) {
+            self.inner.push(signer_id);
+        }
+    }
+
+    pub fn revoke(&mut self, signer_id: &ICSignerId) {
+        if let Some(pos) = self.inner.iter().position(|x| x == signer_id) {
+            self.inner.remove(pos);
+        }
```
Comment (miraclx) [inline contracts/icp/context-config/src/guard.rs]: so this can be `self.inner.remove`

File: contracts/icp/context-config/src/mutate.rs [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+                author_id,
+                application,
+            } => {
+                add_context(
+                    &request.signer_id,
+                    context_id.clone(),
+                    author_id.clone(),
+                    application.clone(),
+                )?;
+                Ok(())
```
Comment (miraclx) [inline contracts/icp/context-config/src/mutate.rs]:
```suggestion
                )
```

and others

File: contracts/icp/context-config/src/mutate.rs [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+}
+
+fn add_context(
+    signer_id: &ICSignerId,
+    context_id: ICContextId,
+    author_id: ICContextIdentity,
+    application: ICApplication,
+) -> Result<(), String> {
+    // 1. Verify signer is the context itself - direct array comparison
+    if signer_id.0 != context_id.0 {
```
Comment (miraclx) [inline contracts/icp/context-config/src/mutate.rs]: `.as_bytes()`

File: contracts/icp/context-config/src/types.rs [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+use candid::CandidType;
+use ed25519_dalek::{Verifier, VerifyingKey};
+use serde::{Deserialize, Serialize};
+
+use crate::guard::Guard;
+
+#[derive(
+    CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Copy, Hash,
+)]
+pub struct Identity(pub [u8; 32]);
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]: these (shouldn't) need to be pub

```suggestion
pub struct Identity([u8; 32]);
```

you can check to be sure, but anything that needs to access the contained data should call `.as_bytes` instead

File: contracts/icp/context-config/src/mutate.rs [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+        RequestKind::Context(ContextRequest { context_id, kind }) => match kind {
+            ContextRequestKind::Add {
+                author_id,
+                application,
+            } => {
+                add_context(
+                    &request.signer_id,
+                    context_id.clone(),
+                    author_id.clone(),
+                    application.clone(),
```
Comment (miraclx) [inline contracts/icp/context-config/src/mutate.rs]: there should be no need to clone here

```suggestion
    match request.kind {
        RequestKind::Context(ContextRequest { context_id, kind }) => match kind {
            ContextRequestKind::Add {
                author_id,
                application,
            } => {
                add_context(
                    &request.signer_id,
                    context_id,
                    author_id,
                    application,
```

File: contracts/icp/context-config/src/types.rs [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+        // Get the signer's public key from the payload
+        let signer_id = f(&self.payload);
+
+        // Convert signer_id to VerifyingKey (public key)
+        let verifying_key =
+            VerifyingKey::from_bytes(&signer_id.0 .0).map_err(|_| "invalid public key")?;
+
+        // Serialize the payload to JSON for verification
+        let message =
+            serde_json::to_vec(&self.payload).map_err(|_| "failed to serialize payload")?;
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]: hm.. why are we using json?

File: contracts/icp/context-config/src/mutate.rs [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+        let mut configs = configs.borrow_mut();
+
+        // Create context with guards
+        let context = Context {
+            application: Guard::new(author_id.clone(), application),
+            members: Guard::new(author_id.clone(), vec![author_id.clone()]),
+            proxy: Guard::new(
+                author_id,
+                format!("{}.{}", configs.next_proxy_id, ic_cdk::api::id()),
+            ),
```
Comment (miraclx) [inline contracts/icp/context-config/src/mutate.rs]: what are ICP addresses like? is there a character limit?

File: contracts/icp/context-config/src/types.rs [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+}
+
+impl Default for ContextConfigs {
+    fn default() -> Self {
+        Self {
+            contexts: HashMap::new(),
+            next_proxy_id: 0,
+        }
+    }
+}
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]: why not put these in `lib.rs`?

File: contracts/icp/context-config/src/types.rs [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+
+impl From<Application<'_>> for ICApplication {
+    fn from(value: Application) -> Self {
+        ICApplication {
+            id: value.id.rt().expect("infallible conversion"),
+            blob: value.blob.rt().expect("infallible conversion"),
+            size: value.size,
+            source: value.source.0.into_owned(),
+            metadata: value.metadata.0.into_inner().into_owned().to_vec(),
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]: this shouldn't be necessary

```suggestion
            metadata: value.metadata.0.into_inner().into_owned(),
```

File: contracts/icp/context-config/src/types.rs [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+    }
+}
+
+impl ReprBytes for ICApplicationId {
+    type EncodeBytes<'a> = [u8; 32];
+    type DecodeBytes = [u8; 32];
+    type Error = LengthMismatch;
+
+    fn as_bytes(&self) -> Self::EncodeBytes<'_> {
+        self.0 .0
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]:
```suggestion
        self.0.as_bytes()
```

and others

File: contracts/icp/context-config/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+#[ic_cdk::init]
+fn init() {
+    CONTEXT_CONFIGS.with(|state| {
+        *state.borrow_mut() = ContextConfigs::default();
+    });
+}
+
+ic_cdk::export_candid!();
```
Comment (miraclx) [inline contracts/icp/context-config/src/lib.rs:55]: vvvery curious about where the state is deserialized

especially since `ContextConfigs` doesn't impl `CandidType`, I can't reason about this

File: contracts/icp/context-config/src/types.rs [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+    }
+
+    #[cfg(test)]
+    pub fn new_with_time(signer_id: ICSignerId, kind: RequestKind, timestamp_ms: u64) -> Self {
+        Self {
+            signer_id,
+            kind,
+            timestamp_ms,
+        }
+    }
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]: why do we have 2 of these?

Comment (miraclx) [reply contracts/icp/context-config/src/types.rs]: Ideally we shouldn't have special behaviour for tests, so it more precisely reflects production behaviour

Comment (alenmestrov) [reply contracts/icp/context-config/src/types.rs]: I had some issues with time generation but this is legacy. 
Removed it from code

File: contracts/icp/context-config/src/types.rs [hunk 18]
```diff
@@ -299,17 +309,21 @@ impl<T: CandidType + Serialize + Clone> ICPSigned<T> {
...
+        })
     }
 
     pub fn parse<R, F>(&self, f: F) -> Result<T, ICPSignedError<R::Error>>
     where
         R: IntoResult<ICSignerId>,
         F: FnOnce(&T) -> R,
     {
-        let bytes = serde_json::to_vec(&self.payload)?;
+        let parsed: T = serde_json::from_slice(&self.payload)?;
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]: `decode_one`? can't guarantee `serde_json` would work

File: contracts/icp/context-config/tests/integration.rs [hunk 19]
```diff
@@ -27,16 +27,10 @@ fn setup() -> (PocketIc, Principal) {
...
-
-    // Sign the serialized message
-    let signature = signer_key.sign(&message);
-
-    ICPSigned {
-        payload: request,
-        signature: signature.to_vec(),
-    }
+    ICPSigned::new(request, |bytes| {
+        Ok::<_, std::convert::Infallible>(signer_key.sign(bytes).to_vec())
```
Comment (miraclx) [inline contracts/icp/context-config/tests/integration.rs]: this?

```suggestion
        signer_key.sign(bytes)
```

File: contracts/icp/context-config/src/mutate.rs [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+            }
+            ContextRequestKind::Grant { capabilities } => grant(
+                &request.signer_id,
+                &context_id.clone(),
+                capabilities.clone(),
+            ),
+            ContextRequestKind::Revoke { capabilities } => revoke(
+                &request.signer_id,
+                &context_id.clone(),
+                capabilities.clone(),
```
Comment (miraclx) [inline contracts/icp/context-config/src/mutate.rs]: remove `.clone()` here, since `Signed::parse` no longer returns `&T`

File: contracts/icp/context-config/src/mutate.rs [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+            proxy: Guard::new(
+                author_id.rt().expect("infallible conversion"),
+                format!("{}.{}", configs.next_proxy_id, ic_cdk::api::id()),
+            ),
+        };
+
+        // Store context
+        if configs
+            .contexts
+            .insert(context_id.clone(), context)
```
Comment (miraclx) [inline contracts/icp/context-config/src/mutate.rs]:
```suggestion
            .insert(context_id, context)
```

File: contracts/icp/context-config/src/mutate.rs [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+
+        // Get mutable access to the application through the Guard
+        let guard_ref = context
+            .application
+            .get(signer_id)
+            .map_err(|e| e.to_string())?;
+        let mut app_ref = guard_ref.get_mut();
+
+        // Replace the application with the new one
+        *app_ref = application.clone();
```
Comment (miraclx) [inline contracts/icp/context-config/src/mutate.rs]:
```suggestion
        *app_ref = application;
```

**Review by miraclx — APPROVED** [review]

File: contracts/icp/context-config/src/types.rs [hunk 23]
```diff
@@ -0,0 +1,360 @@
...
+        S: candid::types::Serializer,
+    {
+        serializer.serialize_null(())
+    }
+}
+
+#[derive(CandidType, Deserialize, Debug, Clone)]
+pub struct ICPSigned<T> {
+    pub payload: Vec<u8>,
+    pub signature: Vec<u8>,
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]:
```suggestion
    payload: Vec<u8>,
    signature: [u8; 32],
```

File: contracts/icp/context-config/src/types.rs [hunk 24]
```diff
@@ -0,0 +1,360 @@
...
+    fn from_bytes<F>(f: F) -> repr::Result<Self, Self::Error>
+    where
+        F: FnOnce(&mut Self::DecodeBytes) -> Bs58Result<usize>,
+    {
+        Identity::from_bytes(f).map(Self)
+    }
+}
+
+#[derive(CandidType, Serialize, Deserialize, Clone, Debug, Hash, Eq, PartialEq)]
+pub struct ICContextId(pub Identity);
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]:
```suggestion
pub struct ICContextId(Identity);
```

File: contracts/icp/context-config/src/types.rs [hunk 25]
```diff
@@ -0,0 +1,360 @@
...
+    fn from_bytes<F>(f: F) -> repr::Result<Self, Self::Error>
+    where
+        F: FnOnce(&mut Self::DecodeBytes) -> Bs58Result<usize>,
+    {
+        Identity::from_bytes(f).map(Self)
+    }
+}
+
+#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct ICApplicationId(pub Identity);
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]:
```suggestion
pub struct ICApplicationId(Identity);
```

File: contracts/icp/context-config/src/types.rs [hunk 26]
```diff
@@ -0,0 +1,360 @@
...
+    fn from_bytes<F>(f: F) -> repr::Result<Self, Self::Error>
+    where
+        F: FnOnce(&mut Self::DecodeBytes) -> Bs58Result<usize>,
+    {
+        Identity::from_bytes(f).map(Self)
+    }
+}
+
+#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct ICBlobId(pub Identity);
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]:
```suggestion
pub struct ICBlobId(Identity);
```

File: contracts/icp/context-config/src/types.rs [hunk 27]
```diff
@@ -0,0 +1,371 @@
...
+        let bytes = candid::encode_one(payload)
+            .map_err(|e| ICPSignedError::SerializationError(e.to_string()))?;
+
+        let signature = sign(&bytes)
+            .into_result()
+            .map_err(ICPSignedError::DerivationError)?;
+
+        Ok(Self {
+            payload: bytes,
+            signature: signature.to_vec(),
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs:319]:
```suggestion
            signature: signature.to_bytes(),
```

File: contracts/icp/context-config/src/types.rs [hunk 28]
```diff
@@ -0,0 +1,371 @@
...
+
+        let key = signer_id
+            .rt::<VerifyingKey>()
+            .map_err(|_| ICPSignedError::InvalidPublicKey)?;
+
+        let signature_bytes: [u8; 64] =
+            self.signature.as_slice().try_into().map_err(|_| {
+                ICPSignedError::SignatureError(ed25519_dalek::ed25519::Error::new())
+            })?;
+        let signature = ed25519_dalek::Signature::from_bytes(&signature_bytes);
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs:344]:
```suggestion
        let signature = ed25519_dalek::Signature::from_bytes(&self.signature);
```

**Review by miraclx — APPROVED** [review]

File: contracts/icp/context-config/src/types.rs (hunk 23 above)
Comment (alenmestrov) [reply contracts/icp/context-config/src/types.rs]: I tried changing signature to [u8; 32] but ed25519 signatures always have 64 bytes so it throwed error when trying to parse it... 
When I switched to [u8; 64] type it gives error because of deseralization...

Comment (miraclx) [reply contracts/icp/context-config/src/types.rs]: > I tried changing signature to [u8; 32] but ed25519 signatures 

sorry, I meant `[u8; 64]`

> When I switched to [u8; 64] type it gives error because of deseralization

what was the error?

File: contracts/icp/context-config/src/types.rs [hunk 29]
```diff
@@ -91,12 +91,16 @@ impl ReprBytes for ICContextIdentity {
...
+pub struct ICContextId(Identity);
 
 impl ICContextId {
     pub fn new(bytes: [u8; 32]) -> Self {
         Self(Identity(bytes))
     }
+
+    pub fn as_bytes(&self) -> [u8; 32] {
+        self.0.as_bytes()
+    }
```
Comment (miraclx) [inline contracts/icp/context-config/src/types.rs]: but you already have this, through `ReprBytes`

**Review by miraclx — APPROVED** [review]
